### ISSUE TITLE
Proposal: No-ASI mode

### ISSUE THREAD

**Issue by manisenkov** [tc39/ecma262#1064]
Inspired by debates at this [PR](https://github.com/tc39/ecma262/pull/1062).

Does it make sense to introduce No-ASI mode that can be activated, for instance, by `"use noasi";` (similarily as `"use strict";` for the strict mode) clause at the begin of the script?

At first look it could be simple to implement this mode in JS interpreters: whenever a case where normally semicolon must be automatically inserted is encountered, it just raise `SyntaxError` instead.

For instance:
```javascript
"use noasi";

function foo () {
    return    // SyntaxError: semicolon is absent in "no-ASI" mode
        42;
}
```

**Comment by WebReflection** [comment]
I think it's been stated various times directives, such as `'use strict'`, won't ever be introduced again in the specs and this seems to be like a good fit for a linter (a new kind, if you want/need).

However, I like the SpiderMonkey behavior where you can (and you should) enable warnings and ASI mistakes will most likely come up (it's also funny the ASI, non-ASI debate always ends up with code nobody writes like the one in your example)

![screenshot from 2018-01-12 14-49-56](https://user-images.githubusercontent.com/85749/34884424-9bf83cf8-f7b4-11e7-9862-4a51daa90d71.png)

TL;DR I think _unreachable code_ warnings are a very good red flag for any production code so that, beside non-ASI gotchas, it'd be cool of engines would all offer similar SpiderMonkey functionality during syntax parsing/code analysis.

**Comment by erights** [comment]
> I think it's been stated various times directives, such as 'use strict', won't ever be introduced again in the specs 

@WebReflection There is no such absolute prohibition. Any proposal to add another would need to meet a very high bar. Indeed, a high enough bar that none is likely. However, if a proposal with a strong enough case is made, we would consider it.

**Comment by annevk** [comment]
This should probably be closed given that feature requests are not supposed to be made in this repository. Please see https://github.com/tc39/ecma262/blob/master/CONTRIBUTING.md for guidelines on how to propose a new feature.
